Multisig/UI/Settings/Passcode/PasscodeSettingsViewController.swift

```diff
@@ -80,42 +86,105 @@ class PasscodeSettingsViewController: UITableViewController {
     }
 
     private func deletePasscode() {
-        let vc = EnterPasscodeViewController()
-        let nav = UINavigationController(rootViewController: vc)
+        withPasscodeAuthentication(for: "Enter Passcode") { [unowned self] success, _, finish in
+            if success {
+                disablePasscode()
+            }
+            finish()
+        }
+    }
 
-        vc.completion = { [weak nav, unowned self] success in
+    private func disablePasscode() {
+        do {
+            try App.shared.auth.deletePasscode()
+            App.shared.snackbar.show(message: "Passcode disabled")
+        } catch {
+            let uiError = GSError.error(
+                description: "Failed to delete passcode",
+                error: GSError.GenericPasscodeError(reason: error.localizedDescription))
+            App.shared.snackbar.show(error: uiError)
+        }
+    }
+
+    private func changePasscode() {
+        withPasscodeAuthentication(for: "Change Passcode", tracking: .changePasscode) { success, nav, finish in
             if success {
+                let changeVC = ChangePasscodeEnterNewViewController {
+                    finish()
+                }
+                nav?.pushViewController(changeVC, animated: true)
+            } else {
+                finish()
+            }
+        }
+    }
+
+    private func toggleUsage(option: PasscodeOptions) {
+        withPasscodeAuthentication(for: "Require to open app") { [unowned self] success, _, finish in
+            if success && AppSettings.passcodeOptions.contains(option) {
+                AppSettings.passcodeOptions.remove(option)
+            } else if success {
+                AppSettings.passcodeOptions.insert(option)
+            }
+
+            if AppSettings.passcodeOptions.isDisjoint(with: [.useForConfirmation, .useForLogin]) {
+                disablePasscode()
+            }
+
+            finish()
+        }
+    }
+
+    private func toggleBiometrics() {
+        withPasscodeAuthentication(for: "Login with biometrics") { success, nav, finish in
+            if AppSettings.passcodeOptions.contains(.useBiometry) {
+                AppSettings.passcodeOptions.remove(.useBiometry)
```

Comment:
I remember we discussed it some time ago, but I have not found it in Passcode v2 requirements is that if action does not reduce security, it should not require confirmation with passcode. Disabling biometry actually increases security, so I would suggest not to ask confirmation for this case. I will add it to the refinement meeting agenda to discuss

Reply:
weren't you saying before that we should request biometry in any case?

Reply:
2nd